safe_transaction_service/contracts/tx_decoder.py

```diff
@@ -452,11 +527,49 @@ class DbTxDecoder(TxDecoder):
     Decode contracts from ABIs in database
     """
 
-    def get_supported_abis(self) -> List[Type[Contract]]:
+    cache_abis_by_address = TTLCache(maxsize=2048, ttl=60 * 5)  # 5 minutes of caching
+
+    @cachedmethod(cache=operator.attrgetter("cache_abis_by_address"))
+    def get_contract_abi(
+        self, address: ChecksumAddress
+    ) -> Optional[Dict[bytes, ABIFunction]]:
+        """
+        :param address: Contract address
+        :return: Dictionary of function selects with ABIFunction if found, `None` otherwise
+        """
+        abis = ContractAbi.objects.filter(contracts__address=address).values_list(
+            "abi", flat=True
+        )
+        if abis:
+            return self._generate_selectors_with_abis_from_abi(abis[0])
+
+    def get_abi_function(
+        self, data: bytes, address: Optional[ChecksumAddress] = None
+    ) -> Optional[ABIFunction]:
+        """
+        :param data: transaction data
+        :param address: contract address in case of ABI colliding
+        :return: Abi function for data if it can be decoded, `None` if not found
+        """
+        selector = data[:4]
+        # Check first that selector is supported on our database
+        if selector in self.fn_selectors_with_abis:
+            # Try to use specific ABI if address provided
+            if address:
+                contract_selectors_with_abis = self.get_contract_abi(address)
```

Comment:
Maybe a small comment here: `If the selector is available in the abi specific for the address we will use that one, otherwise we fallback to the general abi that matches the selector`

Reply:
Agree